Add strict mode support to declare

msssk opened the pull request:
This is an old issue, with [old solutions](http://comments.gmane.org/gmane.comp.web.dojo.devel/16739). Is there a reason it hasn't been addressed? It does require an API change, but the new API is optional - old code does not need to be updated.

RoyTinker commented:
In my opinion, this PR is a more robust change than the one mentioned in http://comments.gmane.org/gmane.comp.web.dojo.devel/16739, because:
- The only API change is that an optional argument was added, so all existing calls to `inherited` can be updated to strict mode by _only_ adding the caller before `arguments`. The other method actually adds 1 more possible type to both the second and third arguments (or an overload, not sure how that could be documented with the existing system) -- which in some cases would require more changes than just adding an argument to the call.
- strict mode calls that fail to pass the caller will give a helpful error message instead of a confusing TypeError

RoyTinker commented:
@msssk Looks like there's a mistake -- please remove the `else` from line 104. That breaks the optionality of arguments 1 and 2.

RoyTinker commented on _base/declare.js:
```diff
@@ -90,23 +90,40 @@ define(["./kernel", "../has", "./lang"], function(dojo, has, lang){
 			a = f;
+			f = g;
+		}
+		else if(typeof args === "function"){
```
change line 104 to `if(typeof args === "function"){`

msssk replied:
Sorry about that - good catch, thanks!

RoyTinker replied:
Great, thanks.

RoyTinker commented:
Looks like when this was last discussed (circa 2012) it wasn't implemented because of known JScript bugs (up through IE8) with named function expressions. That consideration shouldn't preclude this fix -- dojo users don't _have_ to use strict mode, NFEs, etc.

It might be worth mentioning in the documentation that since strict mode support requires use of named function expressions, users supporting IE<=8 should avoid NFEs (and therefore strict mode in classes created using `declare`) because of known JScript bugs.

wkeese commented:
Glad you are adding support for that.   Shouldn't there be a unit test for the new API though?  I assume there's already a test for the old API.

RoyTinker commented:
@msssk @wkeese I attached a suite of 13 unit tests in doh for the new API to the original sitepen ticket.

dylans commented:
@RoyTinker yes, the main issue with that is that we moved from DOH to Intern so we'll need to convert the tests. That said, it sounds like everyone is fine with landing this once we get the tests updated.

msssk commented:
This PR now includes [unit tests for strict mode](https://github.com/dojo/dojo/pull/249/commits/8b65674d1e627dedd853e50725b8e49f5c491416).

The pre-existing `declare` tests continue to pass, indicating that it is very unlikely the new code will affect existing usage of `declare`.

I copied the pre-existing `declare` tests that fail in strict mode and updated them to use the new syntax, and also added a few extra tests to exercise the changes introduced in this PR.

dylans commented on tests/unit/_base/all.js:
```diff
@@ -4,6 +4,7 @@ define([
 	'./array',
 	'./connect',
 	'./declare',
+	'./declare-strict',
```
Should we restrict these to not run in IE8 using `dojo/has`?

msssk replied:
Tests should run in all environments - when people write code using the new syntax that is safe for strict mode environments, we also want to make sure that code works in environments that don't support strict mode.

The only change is a few tests that check for errors being thrown - if strict mode is supported, incorrect syntax results in an error, and we test for that. If strict mode is not supported, no error is thrown (the code works without issue), so I changed the test expectations depending on whether or not the environment supports strict mode.

RoyTinker replied:
They shouldn't run in IE8, but not because it lacks strict mode support. (IE9 also lacks it, but they're fine to run there -- in fact they should.)

The reason is because IE8 and previous has significant JScript bugs with named function expressions, which may invalidate this set of tests.

RoyTinker replied:
@msssk Code targeting IE8 and previous should not use the new syntax as it may break due to JScript bugs NFE hoisting.

msssk replied:
@RoyTinker Thanks, I updated the affected tests to skip running in old IE.

dylans commented:
Closed via b1fc2a2. Thanks @RoyTinker and @msssk for the help!
